### PR TITLE
Introduce flaky tests statistics 

### LINKED ISSUE #358: Print statistics on which tests were restarted and how many times
It will be useful to print statistics on which tests were restarted and how many times. For example, it is convenient to see in real time whether a flaky test was restarted or not after a fix.

### PR DESCRIPTION
**Pull request by ylobankov**
* Consider a test as 'flaky' if it was restarted at least 1 time
  and eventually passed. Print the number of flaked tests and the
  corresponding info in statistics.

* Collect logs of such tests into the 'artifacts' directory, along
  with logs of failed tests.

Example:

    $ ./test-run.py --force replication-luatest/
    Started ./test-run.py --force replication-luatest/
    Running in parallel with 16 workers

    ... <other tests>

    [016] replication-luatest/gh_6033_box_promote_demote>                 [ pass ]
    [014] replication-luatest/gh_5418_qsync_with_anon_te>                 [ pass ]
    [012] replication-luatest/quorum_misc_test.lua                        [ pass ]
    [005] replication-luatest/linearizable_test.lua                       [ fail ]

    ... <traceback>

    [005] Test "replication-luatest/linearizable_test.lua", conf: "None"
    [005] 	failed, rerunning ...
    [005] replication-luatest/linearizable_test.lua                       [ pass ]

    ---------------------------------------------------------------------------------
    Top 10 longest tests (seconds):
    *  67.86 replication-luatest/quorum_orphan_test.lua
    *  26.00 replication-luatest/quorum_misc_test.lua
    *  15.59 replication-luatest/gh_5418_qsync_with_anon_test.lua
    *  13.39 replication-luatest/election_fencing_test.lua
    *   9.77 replication-luatest/gh_5295_split_brain_test.lua
    *   6.95 replication-luatest/no_quorum_test.lua
    *   6.29 replication-luatest/bootstrap_strategy_auto_test.lua
    *   6.26 replication-luatest/gh_5568_read_only_reason_test.lua
    *   5.31 replication-luatest/gh_7253_election_long_wal_write_test.lua
    *   5.13 replication-luatest/gh_7086_box_issue_promote_assert_test.lua
    ---------------------------------------------------------------------------------
    Statistics:
    * pass: 37 (flaky: 1)
    Flaked tasks:
    - [replication-luatest/linearizable_test.lua, null]
    # logfile:        /private/tmp/t/log/005_replication-luatest.log
    # reproduce file: /private/tmp/t/reproduce/005_replication-luatest.list.yaml

    $ ls /tmp/t/artifacts/
    005_replication-luatest/ log/                     reproduce/               statistics/

Closes #358
Part of tarantool/tarantool-qa#299

### CODE REVIEW

File: lib/worker.py [hunk 1]
```diff
@@ -371,7 +373,13 @@ def run_loop(self, task_queue, result_queue):
                 short_status, duration = self.run_task(task_id)
                 retries_left = retries_left - 1
 
-            result_queue.put(self.wrap_result(task_id, short_status, duration))
+            if fails_count > 0 and short_status == 'pass':
+                result_queue.put(self.wrap_result(task_id, 'flaky', duration))
+            else:
+                result_queue.put(
+                    self.wrap_result(task_id, short_status, duration)
+                )
```
Comment (Totktonada) [inline lib/worker.py:381]: I'm in doubt about encoding auxiliary information into the short status. `flaky` here means `passed after one or several fails`. It is `pass` in fact and any code that expects `pass` should also expect `flaky`. It looks error-prone.

We can extend the `WorkerTaskResult` class with one more field or even create a new class to deliver this information to the statistics watcher (so it'll be a separate message). Those ways looks more clean for me.

Comment (ylobankov) [reply lib/worker.py]: Fixed.

File: listeners.py [hunk 2]
```diff
@@ -213,11 +211,12 @@ def __init__(self, get_logfile):
...
-        if not isinstance(obj, WorkerTaskResult):
-            return
+        if isinstance(obj, WorkerTaskResult):
+            if obj.short_status == 'fail' and \
+                    obj.worker_name not in self.failed_workers:
+                self.failed_workers.append(obj.worker_name)
 
-        if obj.short_status in ('fail', 'flaky') and \
-                obj.worker_name not in self.failed_workers:
+        if isinstance(obj, WorkerFlakedTask):
```
Comment (Totktonada) [inline listeners.py:219]: `and obj.worker_name not in self.failed_workers`?

Comment (ylobankov) [reply listeners.py]: Looks like you are right. Fixed.

**Review by Totktonada — APPROVED** [review]
Only glanced briefly: looks okay at high level.

**Pull request merged**